File: examples/zig/tic-tac-toe/TicTacToe_test.zig
const std = @import("std");

const root = @import("../examples_test.zig");
const runar = @import("runar");
const TicTacToe = @import("TicTacToe.runar.zig").TicTacToe;

const contract_source = @embedFile("TicTacToe.runar.zig");

test "compile-check TicTacToe.runar.zig" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    try tmp.dir.writeFile(.{ .sub_path = "TicTacToe.runar.zig", .data = contract_source });

    const path = try std.fs.path.join(allocator, &.{ ".zig-cache", "tmp", tmp.sub_path[0..], "TicTacToe.runar.zig" });
    defer allocator.free(path);
    try root.runar.compileCheckFile(allocator, path);
    try root.runar.compileCheckSource(allocator, contract_source, "TicTacToe.runar.zig");
}

test "tic tac toe join and move update the real contract state" {
    var game = TicTacToe.init(runar.ALICE.pubKey, 100);

    game.join(runar.BOB.pubKey, runar.signTestMessage(runar.BOB));
    try std.testing.expectEqual(@as(i64, 1), game.status);
    try std.testing.expectEqual(@as(i64, 1), game.turn);
    try std.testing.expectEqualSlices(u8, runar.BOB.pubKey, game.playerO);

    game.move(0, runar.ALICE.pubKey, runar.signTestMessage(runar.ALICE));
    try std.testing.expectEqual(@as(i64, 1), game.c0);
    try std.testing.expectEqual(@as(i64, 2), game.turn);

    game.move(4, runar.BOB.pubKey, runar.signTestMessage(runar.BOB));
    try std.testing.expectEqual(@as(i64, 2), game.c4);
    try std.testing.expectEqual(@as(i64, 1), game.turn);
}

test "tic tac toe rejects invalid joins and moves through the real contract" {
    try root.expectAssertFailure("tictactoe-join-twice");
    try root.expectAssertFailure("tictactoe-move-wrong-player");
    try root.expectAssertFailure("tictactoe-move-occupied-cell");
}

test "tic tac toe cancelBeforeJoin validates the real payout output" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    const game = TicTacToe.init(runar.ALICE.pubKey, 100);
    const payout = runar.cat(
        runar.cat(runar.num2bin(100, 8), "1976a914"),
        runar.cat(runar.hash160(runar.ALICE.pubKey), "88ac"),
    );
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{
        .outputHash = runar.hash256(payout),
    }));

    game.cancelBeforeJoin(ctx, runar.signTestMessage(runar.ALICE), runar.hash160(runar.BOB.pubKey), 0);
}

test "tic tac toe cancelBeforeJoin rejects invalid output or signature" {
    try root.expectAssertFailure("tictactoe-cancel-before-join-wrong-sig");
    try root.expectAssertFailure("tictactoe-cancel-before-join-wrong-output");
}

test "tic tac toe cancel validates the real dual-payout output" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    var game = TicTacToe.init(runar.ALICE.pubKey, 100);
    game.join(runar.BOB.pubKey, runar.signTestMessage(runar.BOB));

    const out1 = runar.cat(
        runar.cat(runar.num2bin(100, 8), "1976a914"),
        runar.cat(runar.hash160(runar.ALICE.pubKey), "88ac"),
    );
    const out2 = runar.cat(
        runar.cat(runar.num2bin(100, 8), "1976a914"),
        runar.cat(runar.hash160(runar.BOB.pubKey), "88ac"),
    );
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{
        .outputHash = runar.hash256(runar.cat(out1, out2)),
    }));

    game.cancel(
        ctx,
        runar.signTestMessage(runar.ALICE),
        runar.signTestMessage(runar.BOB),
        runar.hash160(runar.CHARLIE.pubKey),
        0,
    );
}

test "tic tac toe cancel rejects invalid output or signatures" {
    try root.expectAssertFailure("tictactoe-cancel-wrong-sig-x");
    try root.expectAssertFailure("tictactoe-cancel-wrong-sig-o");
    try root.expectAssertFailure("tictactoe-cancel-wrong-output");
}

test "tic tac toe moveAndWin validates the real winner-take-all output" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    const game = TicTacToe{
        .playerX = runar.ALICE.pubKey,
        .betAmount = 100,
        .playerO = runar.BOB.pubKey,
        .c0 = 1,
        .c1 = 1,
        .c3 = 2,
        .c4 = 2,
        .turn = 1,
        .status = 1,
    };

    const payout = runar.cat(
        runar.cat(runar.num2bin(200, 8), "1976a914"),
        runar.cat(runar.hash160(runar.ALICE.pubKey), "88ac"),
    );
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{
        .outputHash = runar.hash256(payout),
    }));

    game.moveAndWin(
        ctx,
        2,
        runar.ALICE.pubKey,
        runar.signTestMessage(runar.ALICE),
        runar.hash160(runar.CHARLIE.pubKey),
        0,
    );
}

test "tic tac toe moveAndWin rejects invalid output or signature" {
    try root.expectAssertFailure("tictactoe-move-and-win-wrong-sig");
    try root.expectAssertFailure("tictactoe-move-and-win-wrong-output");
}

test "tic tac toe moveAndTie validates the real split payout output" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    const game = TicTacToe{
        .playerX = runar.ALICE.pubKey,
        .betAmount = 100,
        .playerO = runar.BOB.pubKey,
        .c0 = 1,
        .c1 = 2,
        .c2 = 1,
        .c3 = 1,
        .c4 = 1,
        .c5 = 2,
        .c6 = 2,
        .c7 = 1,
        .turn = 2,
        .status = 1,
    };

    const out1 = runar.cat(
        runar.cat(runar.num2bin(100, 8), "1976a914"),
        runar.cat(runar.hash160(runar.ALICE.pubKey), "88ac"),
    );
    const out2 = runar.cat(
        runar.cat(runar.num2bin(100, 8), "1976a914"),
        runar.cat(runar.hash160(runar.BOB.pubKey), "88ac"),
    );
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{
        .outputHash = runar.hash256(runar.cat(out1, out2)),
    }));

    game.moveAndTie(
        ctx,
        8,
        runar.BOB.pubKey,
        runar.signTestMessage(runar.BOB),
        runar.hash160(runar.CHARLIE.pubKey),
        0,
    );
}

test "tic tac toe moveAndTie rejects invalid output or signature" {
    try root.expectAssertFailure("tictactoe-move-and-tie-wrong-sig");
    try root.expectAssertFailure("tictactoe-move-and-tie-wrong-output");
}

File: examples/ts/tic-tac-toe/TicTacToe.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createHash } from 'node:crypto';
import { TestContract, ALICE, BOB, signTestMessage } from 'runar-testing';

// ---------------------------------------------------------------------------
// Byte-level helpers for pre-computing terminal method output hashes.
// Used to configure setMockPreimageBytes so hash256(payout) == extractOutputHash.
// ---------------------------------------------------------------------------

function hexToU8(hex: string): Uint8Array {
  const buf = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) buf[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  return buf;
}
function catU8(...arrays: Uint8Array[]): Uint8Array {
  const total = arrays.reduce((s, a) => s + a.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const a of arrays) { out.set(a, off); off += a.length; }
  return out;
}
function num2binU8(v: bigint, len: bigint): Uint8Array {
  const buf = new Uint8Array(Number(len));
  let abs = v < 0n ? -v : v;
  for (let i = 0; i < Number(len) && abs > 0n; i++) { buf[i] = Number(abs & 0xffn); abs >>= 8n; }
  if (v < 0n) buf[Number(len) - 1] |= 0x80;
  return buf;
}
function hash160U8(data: Uint8Array): Uint8Array {
  const sha = createHash('sha256').update(data).digest();
  return new Uint8Array(createHash('ripemd160').update(sha).digest());
}
function hash256U8(data: Uint8Array): Uint8Array {
  const h1 = createHash('sha256').update(data).digest();
  return new Uint8Array(createHash('sha256').update(h1).digest());
}
function p2pkhOutput(satoshis: bigint, playerHex: string, prefixHex: string, suffixHex: string): Uint8Array {
  return catU8(num2binU8(satoshis, 8n), hexToU8(prefixHex), hash160U8(hexToU8(playerHex)), hexToU8(suffixHex));
}

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'TicTacToe.runar.ts'), 'utf8');

const PLAYER_X = ALICE.pubKey;
const PLAYER_O = BOB.pubKey;
const ZERO_PK = '00'.repeat(33);
const SIG_X = signTestMessage(ALICE.privKey);
const SIG_O = signTestMessage(BOB.privKey);
const BET_AMOUNT = 1000n;
const P2PKH_PREFIX = '1976a914';
const P2PKH_SUFFIX = '88ac';

function makeGame(overrides: Record<string, unknown> = {}) {
  return TestContract.fromSource(source, {
    playerX: PLAYER_X,
    betAmount: BET_AMOUNT,
    p2pkhPrefix: P2PKH_PREFIX,
    p2pkhSuffix: P2PKH_SUFFIX,
    playerO: ZERO_PK,
    c0: 0n, c1: 0n, c2: 0n,
    c3: 0n, c4: 0n, c5: 0n,
    c6: 0n, c7: 0n, c8: 0n,
    turn: 0n,
    status: 0n,
    ...overrides,
  });
}

function makePlayingGame(overrides: Record<string, unknown> = {}) {
  return makeGame({
    playerO: PLAYER_O,
    status: 1n,
    turn: 1n,
    ...overrides,
  });
}

describe('TicTacToe', () => {
  describe('join', () => {
    it('allows player O to join a waiting game', () => {
      const game = makeGame();
      const result = game.call('join', { opponentPK: PLAYER_O, sig: SIG_O });
      expect(result.success).toBe(true);
      expect(game.state.playerO).toBe(PLAYER_O);
      expect(game.state.status).toBe(1n);
      expect(game.state.turn).toBe(1n);
    });

    it('rejects join when game is already playing', () => {
      const game = makePlayingGame();
      const result = game.call('join', { opponentPK: PLAYER_O, sig: SIG_O });
      expect(result.success).toBe(false);
    });
  });

  describe('move', () => {
    it('allows player X to place a mark on an empty cell', () => {
      const game = makePlayingGame();
      const result = game.call('move', { position: 0n, player: PLAYER_X, sig: SIG_X });
      expect(result.success).toBe(true);
      expect(game.state.c0).toBe(1n);
      expect(game.state.turn).toBe(2n);
    });

    it('allows player O to place a mark on their turn', () => {
      const game = makePlayingGame({ turn: 2n });
      const result = game.call('move', { position: 4n, player: PLAYER_O, sig: SIG_O });
      expect(result.success).toBe(true);
      expect(game.state.c4).toBe(2n);
      expect(game.state.turn).toBe(1n);
    });

    it('rejects move on an occupied cell', () => {
      const game = makePlayingGame({ c0: 1n });
      const result = game.call('move', { position: 0n, player: PLAYER_X, sig: SIG_X });
      expect(result.success).toBe(false);
    });

    it('rejects move when game is not playing', () => {
      const game = makeGame();
      const result = game.call('move', { position: 0n, player: PLAYER_X, sig: SIG_X });
      expect(result.success).toBe(false);
    });

    it('tracks state across multiple moves', () => {
      const game = makePlayingGame();
      game.call('move', { position: 0n, player: PLAYER_X, sig: SIG_X });
      expect(game.state.c0).toBe(1n);
      expect(game.state.turn).toBe(2n);

      game.call('move', { position: 4n, player: PLAYER_O, sig: SIG_O });
      expect(game.state.c4).toBe(2n);
      expect(game.state.turn).toBe(1n);

      game.call('move', { position: 8n, player: PLAYER_X, sig: SIG_X });
      expect(game.state.c8).toBe(1n);
      expect(game.state.turn).toBe(2n);
    });
  });

  describe('moveAndWin', () => {
    it('succeeds when the move completes a row win', () => {
      const game = makePlayingGame({
        c0: 1n, c1: 1n,
        c3: 2n, c4: 2n,
        turn: 1n,
      });
      // Pre-compute the payout hash so extractOutputHash returns it.
      const payout = p2pkhOutput(BET_AMOUNT * 2n, PLAYER_X, P2PKH_PREFIX, P2PKH_SUFFIX);
      game.setMockPreimageBytes({ outputHash: hash256U8(payout) });
      const result = game.call('moveAndWin', { position: 2n, player: PLAYER_X, sig: SIG_X, changePKH: '00', changeAmount: 0n });
      if (!result.success) console.error('moveAndWin error:', result.error);
      expect(result.success).toBe(true);
    });
  });

  describe('moveAndTie', () => {
    it('succeeds when the board becomes full with no winner', () => {
      const game = makePlayingGame({
        c0: 1n, c1: 2n, c2: 1n,
        c3: 1n, c4: 1n, c5: 2n,
        c6: 2n, c7: 1n,
        turn: 2n,
      });
      // Pre-compute the tie payout hash (split between both players).
      const out1 = p2pkhOutput(BET_AMOUNT, PLAYER_X, P2PKH_PREFIX, P2PKH_SUFFIX);
      const out2 = p2pkhOutput(BET_AMOUNT, PLAYER_O, P2PKH_PREFIX, P2PKH_SUFFIX);
      game.setMockPreimageBytes({ outputHash: hash256U8(catU8(out1, out2)) });
      const result = game.call('moveAndTie', { position: 8n, player: PLAYER_O, sig: SIG_O, changePKH: '00', changeAmount: 0n });
      expect(result.success).toBe(true);
    });
  });

  describe('full game flow', () => {
    it('plays through join + moves + X wins top row', () => {
      const game = makeGame();
      const joinResult = game.call('join', { opponentPK: PLAYER_O, sig: SIG_O });
      expect(joinResult.success).toBe(true);
      expect(game.state.status).toBe(1n);

      // X@0, O@3, X@1, O@4 — set up X to win with position 2 (top row)
      game.call('move', { position: 0n, player: PLAYER_X, sig: SIG_X });
      expect(game.state.c0).toBe(1n);

      game.call('move', { position: 3n, player: PLAYER_O, sig: SIG_O });
      expect(game.state.c3).toBe(2n);

      game.call('move', { position: 1n, player: PLAYER_X, sig: SIG_X });
      expect(game.state.c1).toBe(1n);

      game.call('move', { position: 4n, player: PLAYER_O, sig: SIG_O });
      expect(game.state.c4).toBe(2n);
      expect(game.state.turn).toBe(1n); // X's turn

      // X plays position 2 to win top row (0,1,2).
      // Pre-compute payout hash so extractOutputHash returns the right value.
      const payout = p2pkhOutput(BET_AMOUNT * 2n, PLAYER_X, P2PKH_PREFIX, P2PKH_SUFFIX);
      game.setMockPreimageBytes({ outputHash: hash256U8(payout) });
      const winResult = game.call('moveAndWin', { position: 2n, player: PLAYER_X, sig: SIG_X, changePKH: '00', changeAmount: 0n });
      expect(winResult.success).toBe(true);
    });
  });
});
